### ISSUE TITLE
How to set environment variables?

### ISSUE THREAD

**Issue by kpgarrod** [ngbp/ngbp#165]
I am trying to deploy an app to a staging/production environment. I need to tell it to use a different database from the development env. Surely this is a requirement for any non-trivial Angular app, but I don't see a way of passing a variable into the static (compiled) files.

I happen to be using a Node.js server, but I think the issue will be the same with any backend.

Any suggestions welcome!

**Comment by joshdmiller** [comment]
I don't think this is really an ngBoilerplate question. The database isn't accessed by the client, which is all ngBoilerplate ever touches. What specifically are you trying to do?

**Comment by kpgarrod** [comment]
Sorry I didn't describe the issue clearly enough. 

In my particular case, the client IS accessing the database (Firebase). It is also accessing my own backend api server and other third-party services. Isn't that typical for non-trivial Angular apps?

Their are really three related issues:
1. As much as I like the idea, I can't practically run the app from the src directory (with 'file://'), because calls to the external services don't work. This is easily solved by adding a server, as you pointed out in another post. So, I have added an Express server to serve build/index.html, which is effectively my 'development' environment. This version of the app accesses my development Firebase and my development api server, which all works fine.
2. Now I want to deploy the app to a 'staging' environment, which looks more like production - in my case Heroku. So, I Grunt compile the app and I have the compiled version in the 'bin' directory. I can then serve the compiled version using an Express server. The problem is that there is no way that I could find of pointing this version of the app at different backend (Firebase, api etc.) servers. I can't add a config script to the bin directory, because it gets overwritten by the compile. I can't access the 'NODE_ENV' constant from the Angular app. I can't include an ejs variable in the html page because the page is rendered by the compile process and ejs is resolved there. Apart from urls for backend services, there are things like OAuth secrets and so on which have to be configured differently for different environments.
3. When I have overcome issue 2, I still only have a 'staging' version of the app. In order to deploy to production, I need to set new values for the backend server addresses etc. I realise that the development/staging/production model is not universal, but I think it is usual for non-trivial apps.

I have hacked a solution to these issues with some help from Martin Gontovnikas. He will be submitting a pull request. The idea is to:
1. Add some environment configuration to build.config.js.
2. Create a compile_dir for each environment defined in the environment config.
3. Include a config script tag into index.html. 
4. Generate a config.js file for each environment using the environment config data from build.config.js.

It's not beautiful and it needs some generalising but it should get me unstuck at least!

**Comment by joshdmiller** [comment]
> In my particular case, the client IS accessing the database (Firebase). It is also accessing my own backend api server and other third-party services. Isn't that typical for non-trivial Angular apps?

Some third-party mashups are common for large apps, sure - but they're typically social-type stuff and not the actual backend. I don't really consider Firebase to be a "database" in the traditional sense, but rather an outsourced, half-PaaS/half-SaaS backend for web apps.

> As much as I like the idea, I can't practically run the app from the src directory (with 'file://'), because calls to the external services don't work.

The idea isn't to **run** the app from `file://` but to **develop** the app from `file://`. In development, API calls should be mocked out - including the ones to third-party services. I recognize that many developers find that inconvenient, but it is arguably the best practices approach. As for Firebase, we should consider it more like our server-side API than like our database - and we should treat it as such. That is, it should be mocked out as well during testing and development.

> I have hacked a solution to these issues with some help from Martin Gontovnikas. He will be submitting a pull request. The idea is to:
> 1. Add some environment configuration to build.config.js.
> 2. Create a compile_dir for each environment defined in the environment config.
> 3. Include a config script tag into index.html.
> 4. Generate a config.js file for each environment using the environment config data from build.config.js.

I think you're probably overcomplicating things here. ngBoilerplate is going to work like any other AngularJS build system. Building three compile targets based on stuff in `build.conf.js`, unless I'm misunderstanding, sounds like overkill. Here's a simpler approach - 

Create a `src/conf/` dir and store three files in there: `staging.js`, `prod.js`, and `dev.js`. The former two contain all the configuration necessary to access the staging and production environments. They're services:

``` js
// prod.js
angular.module( "myApp.conf", [] )
.value( "configuration", {
  apiEndpoint: "...",
  // ...
})
;
```

The last file, `dev.js`, shouldn't be like the above but rather should contain mocks:

``` js
angular.module( "myApp.conf", [ "ngMock" ] )
.config( function( $httpBackend ) {
  // define your mocks, or import them
})
;
```

The idea is that in development, we should be programming _to spec_; if the server and client are developed to the same spec, they'll work when you hit production. If not, they won't. But if you develop them concomitantly by intermingling them, instead of correcting which one doesn't meet spec, we too often simply "make it work", which hurts us in the long run. But anyway, if you really insisted, `dev.js` could simply be like the other two.

Now you have centralized configuration that leverages AngularJS's wonderful DI - all we need to do is use ngbp/grunt to ensure only one is included at a time. So we can just make a couple simple changes:

``` js
// build.config.js
env: "dev",

// ...

app_files: {
  // ...
  js: [ ..., "src/conf/<%= env %>.js", ... ]
},
```

``` js
// Gruntfile.js

// this is the old build task, exactly as it was, just renamed
grunt.task.registerTask( "ngbpbuild", [
  // ...
]);

grunt.task.registerTask( "build", function ( env ) {
  env = env || "dev";
  grunt.config.set( "env", env );
  grunt.task.run( "ngbpbuild" );
});
```

And then a default "build" is simply a dev build, but you can also specify a type to build that kind, but all will still end up in the `bin/` dir:

``` sh
$ grunt build:prod
# or
$ grunt build:staging
```

This is a very simple change, but it's also very clean. There's no need for a separate compile directory for each environment; just make clever use of DI like you should in any other AngularJS app. I can also think of a dozen other ways to solve this problem too if you'd like more options, but none of them are really ngBoilerplate-specific.

Hope this helps!

**Comment by kpgarrod** [comment]
Thank you very much for taking the time to reply. Your solution is certainly cleaner and neater than my clumsy attempt! I made a couple of minor changes though:
1. Change the app_files definition to include  `...'!src/conf/*.js', "src/conf/<%= env %>.js",...` to prevent all config files being gobbled up all the time.
2. Left the original `grunt build` task as it was and added a `grunt deploy` task with the env parameter instead of ngbpbuild:

``` javascript
grunt.task.registerTask( "deploy", function ( env ) {
    env = env || "prod";
    grunt.config.set( "env", env );
    grunt.task.run('default');
  });
```

This works fine. The only downside is that if I `grunt deploy:staging`, I cannot now just push the github staging branch to production when I'm ready. I need to `grunt deploy:prod` first.   

The reason I considered the issue to be ngBoilerplate-specific is that I originally created the app with Yeoman and did not face the same problems. The fact that the compiled app is essentially a static file, rather than a page that is rendered by a server, impacts the deployment process. In my opinion, the project would benefit from this kind of deployment
functionality.

I really appreciate the work that you are doing with ngBoilerplate and your insights into development process best practices. I hope that by raising these issues I am making a constructive contribution.

**Comment by Keats** [comment]
If you want to have only one file, you can use https://npmjs.org/package/grunt-replace and just have a replace task before compiling where you define your settings for each env

**Comment by tony-kerz** [comment]
@kpgarrod take a look at this to see if it helps with the concern you are trying to address:

https://github.com/werk85/grunt-ng-constant

**Comment by fraserxu** [comment]
@kpgarrod `ng-constant` works for me. You may find an example here. https://github.com/Wiredcraft/feathers/blob/master/build.config.js#L41

We're using `ng-constant` to access different api according the environment, and we also have some custom settings that build on top of `ngbp`.
